File: srcs/django-files/backend/tpong/static/tpong/PongBall.js
import { Util } from "./Util.js";
import { vec2 } from "./vec2.js";
import collisionManager from "./CollisionManager.js";
import { Color } from "./Color.js";


export class PongBall
{
	constructor(radius, defpos)
	{
		this.type = "ball"
		this.radiusNormal = radius;
		this.ballNormal = new vec2(0.5, 0.5);
		this.velocityNormal = new vec2()//(0.3, 0.2);
		this.cuid = collisionManager.addCollidableCircle(new vec2(), this.radiusNormal, this.onCollision.bind(this), this);
		this.color = new Color(255, 255, 255, 1)
	}

	//getWindowNormal()
	//{
	//	return new vec2(this.radiusNormal).add(this.ballNormal.multiply(new vec2(1).subtract(new vec2(this.radiusNormal))))
	//}

	syncDOM()
	{
		var pos = this.ballNormal.subtract(new vec2(this.radiusNormal));
		var ball = Util.getBallHtmlObject();
		const boundingBox = Util.getGameWindowHtmlObject().getBoundingClientRect();
		var aspectRatioW = boundingBox.width / boundingBox.height
		
		pos = pos.multiply(new vec2(boundingBox.width, boundingBox.height - ((this.radiusNormal * 2) * boundingBox.height) * (aspectRatioW - 1)));
		//pos = pos.add(new vec2(boundingBox.left + document.documentElement.scrollLeft, boundingBox.top + document.documentElement.scrollTop));
		ball.style.backgroundColor = this.color.toRgbCss()
		ball.style.height = ((this.radiusNormal * 2) * boundingBox.height) * aspectRatioW + 'px';
		ball.style.width = ((this.radiusNormal * 2) * boundingBox.width) + 'px';
		ball.style.position = 'absolute';
		ball.style.left = pos.x + 'px';
		ball.style.top = pos.y + 'px';
	}

	reset()
	{
		this.ballNormal = new vec2(0.5, 0.5);
		this.velocityNormal = new vec2(0.0, 0.0);
	}

	simulate(delta)
	{
		//this.velocityNormal.y += 0.0981 * delta;
		this.ballNormal = this.ballNormal.add(this.velocityNormal.multiply(new vec2(delta)));
		this.ballNormal = this.ballNormal.clamp(new vec2(1 - this.radiusNormal), new vec2(this.radiusNormal));
		collisionManager.updateCollidableObject(this.cuid, this.ballNormal);
		this.color.lerp(new Color(255, 255, 255, 1), delta)
	}

	onCollision(collision, obj)
	{
		if (collision.clipPos.x)
			this.ballNormal.x = collision.clipPos.x //* (1 + this.radiusNormal);
		if (collision.clipPos.y)
			this.ballNormal.y = collision.clipPos.y //* (1 + this.radiusNormal);
		this.ballNormal = this.ballNormal.clamp(new vec2(1 - this.radiusNormal), new vec2(this.radiusNormal));
		if (collision.side == 'left' && this.velocityNormal.x < 0
			|| collision.side == 'right' && this.velocityNormal.x > 0)
			this.velocityNormal.x *= -1;
		if (collision.side == 'top' && this.velocityNormal.y < 0
			|| collision.side == 'bottom' && this.velocityNormal.y > 0)
			this.velocityNormal.y *= -1;
	}
}

File: srcs/django-files/backend/tpong/game/engine/gameObjects/PongBall.py
from tpong.game.engine.Vec2 import Vec2
from .Collidable import Collidable
import logging

logger = logging.getLogger(__name__)

class PongBall(Collidable):
	def __init__(self) -> None:
		super().__init__('circle')
		self.radius = 0.01
		self.position = Vec2(0.5, 0.5)
		self.velocity = Vec2()
		#self.velocity.x = -0.3
		#self.velocity.y = -0.15

	def simulate(self, delta):
		newPos = self.position + (self.velocity * delta)
		newPos.clamp(1 - self.radius, self.radius)
		self.position = newPos
	
	def serializeState(self):
		return {
			'objName': 'pongball',
			'ballPosition': {
				'x': self.position.x,
				'y': self.position.y
			},
			'ballVelocity': {
				'x': self.velocity.x,
				'y': self.velocity.y
			},
			'ballRadius': self.radius
		}
	
	def onCollision(self, collision):
		#logger.warning(f"Ball Collision: x:{collision['side'].x} y:{collision['side'].y}")
		if collision['clipPos'].x:
			self.position.x = collision['clipPos'].x
		if collision['clipPos'].y:
			self.position.y = collision['clipPos'].y
		
		if (collision['side'].x == -1 and self.velocity.x < 0) or (collision['side'].x == 1 and self.velocity.x > 0):
			self.velocity.x *= -1
		if (collision['side'].y == 1 and self.velocity.y < 0) or (collision['side'].y == -1 and self.velocity.y > 0):
			self.velocity.y *= -1

	def reset(self):
		self.position = Vec2(0.5, 0.5)
		self.velocity = Vec2()